modules/standard/Heap.chpl

```diff
@@ -0,0 +1,420 @@
+          if (_greater(_data[greaterChild+1],_data[greaterChild])) {
+            // if the right child is greater, then update the greaterChild
+            greaterChild += 1;
+          }
+        }
+        // if the greaterChild's value is greater than current node, then swap and continue
+        if (_greater(_data[greaterChild],_data[pos])) {
+          _data[greaterChild] <=> _data[pos];
+          pos = greaterChild;
+        }
+        else break;
+      }
+    }
+
+    pragma "no doc"
+    proc _push(in element: eltType)
+    lifetime this < element {
+      _data.append(element);
+      _heapify_up(_data.size-1);
+    }
+    /*
+      Push an element into the heap
+
+      :arg element: The element that will be pushed
+      :type element: `eltType`
+    */
+    proc push(in element: eltType)
+    lifetime this < element {
+      _enter();
+      _push(element);
+      _leave();
+    }
+
+    /*
+      Push elements of a list into the heap
+
+      :arg x: The list of which elements is to push.
+      :type x: `list(eltType)`
+    */
+    proc push(ref x: list(eltType)) {
+      _enter();
+      for e in x do
+        _push(x);
+      _leave();
+    }
+
+    /*
+      Push elements in an array into a heap.
+
+      :arg x: The array of which elements is to push.
+      :type x: `[?d] eltType`
+    */
+    proc push(ref x:[?d] eltType) {
+      _enter();
+      for e in x do
+        _push(e);
+      _leave();
+    }
+
+    /*
+      Pop an element and return it.
+
+      :return: the top element
+      :rtype: eltType
+    */
+    proc pop(): eltType {
+      _enter();
+      if (boundsChecking && isEmpty()) {
+        boundsCheckHalt("Called \"heap.pop\" on an empty heap.");
+      }
+
+      if _data.size != 1 then
+        _data(0) <=> _data(_data.size-1);
+
+      var ret = _data.pop();
+      _heapify_down(0);
+      _leave();
+      return ret;
+    }
+
+    /*
+      Iterate over the elements of this heap
+
+        .. note::
+          Elements are not yielded in order
+
+    */
+    iter these() ref {
+      for e in _data {
+        yield e;
+      }
+    }
+
+    /*
+      Iterate over the elements of this heap
+      
+        .. note::
+          Elements are yield in order
+    */
+    iter consume() {
+      var h = this;
+      while !h.isEmpty() {
+        yield h.pop();
+      }
+    }
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
+      elements contained in this heap.
+
+      :return: A new DefaultRectangular array.
+    */
+    proc const toArray(): [] eltType {
+      if !isCopyableType(eltType) then
+        compilerError("toArray() method is not avaliable on a 'heap'",
+                      " with elements of a type that can't be copied, here: ",
+                      eltType: string);
+      _enter();
+      var l = _data.toArray();
+      _leave();
+      return l;
+    }
+
+    /*
+      Write the contents of this list to a channel.
```

Comment:
This should also note that the output isn't sorted.  And refer to this as a heap.